schematic/src/i_callbacks.c

```diff
@@ -2459,52 +2459,7 @@ DEFINE_I_CALLBACK(attributes_attach)
   o_undo_savestate_old(w_current, UNDO_ALL);
 }
 
-/*! \todo Finish function documentation!!!
- *  \brief
- *  \par Function Description
- *
- */
-DEFINE_I_CALLBACK(attributes_detach)
-{
-  GschemToplevel *w_current = GSCHEM_TOPLEVEL (data);
-  GList *s_current;
-  OBJECT *o_current;
-  GList *detached_attribs = NULL;
-
-  g_return_if_fail (w_current != NULL);
 
-  /* This is a new addition 3/15 to prevent this from executing
-   * inside an action */
-  if (w_current->inside_action) {
-    return;
-  }
-
-  /* same note as above on i_update_middle_button */
-  i_update_middle_button(w_current, i_callback_attributes_detach,
-                         _("Detach"));
-
-  s_current = geda_list_get_glist( gschem_toplevel_get_toplevel (w_current)->page_current->selection_list );
-  while (s_current != NULL) {
-    o_current = (OBJECT *) s_current->data;
-    if (o_current) {
-      if (o_current->attribs) {
-        detached_attribs = g_list_concat (g_list_copy (o_current->attribs),
-                                          detached_attribs);
-        o_attrib_detach_all (gschem_toplevel_get_toplevel (w_current), o_current);
-        gschem_toplevel_get_toplevel (w_current)->page_current->CHANGED=1;
```

Comment:
You've missed setting the page *dirty* flag after detachment in you Scheme code.

Reply:
Thanks! Fixed.
There is a general problem with `set-page-dirty!`: it's a library function, and
we can't determine if it was called in `lepton-schematic`. So, for example, we
cannot update GUI, indicating that the page has been changed.

Reply:
We could make up a Scheme callback just for `lepton-schematic` using *glib* functionality to send an appropriate message to the program.  Likewise `log!` and the stuff around it used for message logging in or GUI and non-GUI programs now.

Reply:
Although logging is implemented using *glib* functions, `lepton-schematic`
receives log messages simply via a global function pointer
(`x_log_update_func`, `s_log.c`).
Page change notifications could be done similarly, but it's not an easy task
with the current code state and it's not enough. Scheme code should also be
notified (if possible, via guile hooks).
`CHANGED` field of `st_page` structure should be made private and modification
status should be accessed via functions and in a centralized manner.
Now there is `gschem_toplevel_page_content_changed()` function.
Comment says that *"This function is temporary until library functions
can emit signals"*. Why temporary? What signals?
All this stuff makes an impression of incomplete, disorganized code;
we need to fix it.

Reply:
It was temporary because other devs didn't expect the events go the way we watch now. Previously, things were worse, AFAICR. Now it's our task to make them better. It was planned that a C function editing a page should send, say, the "changed" signal to the program, and notify the guile side as well.  The development of `geda-gaf` went parallel in two (or even more) separate ways. Naturally, some people prefer C and gobjects (like Edward), others are admiring for Scheme (like me).  The previous code base had neither of those two approaches. Most people just wanted to improve the C code we had because despite of its quirks it's quite usable, isn't it? ;-)